### PR TITLE
Imagemagick: modernize package options

### PR DESCRIPTION
**Pull request by alyssais**
###### Description of changes

Throughout Nixpkgs, we now tend to avoid packages being null outside of bootstrapping, in favour of dedicated options that can have more complex defaults set, e.g. per-platform.  This simplifies and regularizes the imagemagick package definition.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by lucasew — APPROVED** [review]

**Comment by Mindavi** [comment]
Nice cleanup, though I do wonder if we don't want to remove a lot of these options. Do we really want to expose all these settings?

File: pkgs/applications/graphics/ImageMagick/6.x.nix [hunk 1]
```diff
@@ -30,35 +49,47 @@ stdenv.mkDerivation rec {
...
-    ++ lib.optionals (ghostscript != null)
-      [ "--with-gs-font-dir=${ghostscript}/share/ghostscript/fonts"
-        "--with-gslib"
-      ]
-    ++ lib.optionals (stdenv.hostPlatform.isMinGW)
-      [ "--enable-static" "--disable-shared" ] # due to libxml2 being without DLLs ATM
-    ;
+  configureFlags = [
+    "--with-frozenpaths"
+    (lib.withFeatureAs (arch != null) "gcc-arch" arch)
```
Comment (Mindavi) [inline pkgs/applications/graphics/ImageMagick/6.x.nix:54]: Is `--without-gcc-arch` default, or will it automatically do `--without-gcc-arch` with the `lib.withFeatureAs` function?

**Review by Mindavi — APPROVED** [review]

File: pkgs/applications/graphics/ImageMagick/6.x.nix (hunk 1 above)
Comment (alyssais) [reply pkgs/applications/graphics/ImageMagick/6.x.nix:54]: The latter.

File: pkgs/top-level/all-packages.nix [hunk 2]
```diff
@@ -27781,66 +27781,64 @@ with pkgs;
...
+    librsvgSupport = false;
+    libtiffSupport = false;
+    libxml2Support = false;
+    openjpegSupport = false;
+    libwebpSupport = false;
+    libheifSupport = false;
+    libde265Support = false;
   };
 
   imagemagick6 = callPackage ../applications/graphics/ImageMagick/6.x.nix {
```
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:27806]:
```suggestion
  imagemagick6 = callPackage ../applications/graphics/ImageMagick/6.nix {
```

File: pkgs/top-level/all-packages.nix [hunk 3]
```diff
@@ -27781,66 +27781,64 @@ with pkgs;
...
+    libtiffSupport = false;
+    libxml2Support = false;
+    openjpegSupport = false;
+    libwebpSupport = false;
+    libheifSupport = false;
   });
 
-  imagemagick = lowPrio (imagemagickBig.override {
-    ghostscript = null;
+  imagemagick = lowPrio (callPackage ../applications/graphics/ImageMagick/7.0.nix {
```
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:27836]:
```suggestion
  imagemagick = lowPrio (callPackage ../applications/graphics/ImageMagick/7.nix {
```

File: pkgs/applications/graphics/ImageMagick/6.x.nix [hunk 4]
```diff
@@ -30,35 +49,47 @@ stdenv.mkDerivation rec {
...
-  buildInputs =
-    [ zlib fontconfig freetype ghostscript
-      liblqr1 libpng libtiff libxml2 libheif libde265 djvulibre
-    ]
-    ++ lib.optionals (!stdenv.hostPlatform.isMinGW)
-      [ openexr librsvg openjpeg ]
-    ++ lib.optionals stdenv.isDarwin
-      [ ApplicationServices Foundation ];
+  buildInputs = [ ]
+    ++ lib.optional zlibSupport zlib
```
Comment (SuperSandro2000) [inline pkgs/applications/graphics/ImageMagick/6.x.nix:67]:
```suggestion
  buildInputs = lib.optional zlibSupport zlib
```

**Review by SuperSandro2000 — COMMENTED** [review]
good work

File: pkgs/applications/graphics/ImageMagick/6.x.nix (hunk 4 above)
Comment (alyssais) [reply pkgs/applications/graphics/ImageMagick/6.x.nix:67]: I like it as is, for consistency.  I've seen this style in a lot of other places in Nixpkgs.

File: pkgs/top-level/all-packages.nix (hunk 3 above)
Comment (alyssais) [reply pkgs/top-level/all-packages.nix:27836]: Way ahead of you. :) https://github.com/NixOS/nixpkgs/pull/181454

File: pkgs/applications/graphics/ImageMagick/6.x.nix (hunk 4 above)
Comment (SuperSandro2000) [reply pkgs/applications/graphics/ImageMagick/6.x.nix:67]: Whenever I change a line next to such a line I remove it but whatever you prefer.

**Comment by alyssais** [comment]
`imagemagick.passthru.tests` failing because of https://github.com/NixOS/nixpkgs/pull/181776.

**Pull request merged**